weld/expr_hash.rs

```diff
@@ -0,0 +1,189 @@
+
+use super::ast::*;
+use super::ast::ExprKind::*;
+use super::ast::LiteralKind::*;
+use super::error::*;
+
+use std::collections::hash_map::Entry;
+use std::hash::{Hash, Hasher};
+
+use std::fmt;
+
+/// A signature which uniquely represents an Expression in a concise manner.
+pub struct ExprHash {
+    hasher: fnv::FnvHasher,
+}
+
+impl fmt::Debug for ExprHash {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "ExprHash({})", self.hasher.finish())
+    }
+}
+
+impl ExprHash {
+    /// Recurisvely computes signatures for an expression and each of its subexpressions.
+    /// The precise symbol names defined within the expression are ignored.
+    fn from_expr<'a, T: TypeBounds>(&mut self, expr: &'a Expr<T>,
+                                    symbol_positions: &mut fnv::FnvHashMap<&'a Symbol, Vec<i32>>,
+                                    max_id: &mut i32) -> WeldResult<()> {
+        // In expressions that define new symbols, subexpressions will be handled in the match.
+        let mut finished_subexpressions = false;
+        // Hash the type.
+        expr.ty.hash(&mut self.hasher);
+        // Hash the name, which represents the kind of the expression.
+        expr.kind.name().hash(&mut self.hasher);
+        // Hash the kind.
+        match expr.kind {
+            Literal(ref kind) => {
+                match *kind {
+                    BoolLiteral(v) => v.hash(&mut self.hasher),
+                    I8Literal(v) => v.hash(&mut self.hasher),
+                    I16Literal(v) => v.hash(&mut self.hasher),
+                    I32Literal(v) => v.hash(&mut self.hasher),
+                    I64Literal(v) => v.hash(&mut self.hasher),
+                    U8Literal(v) => v.hash(&mut self.hasher),
+                    U16Literal(v) => v.hash(&mut self.hasher),
+                    U32Literal(v) => v.hash(&mut self.hasher),
+                    U64Literal(v) => v.hash(&mut self.hasher),
+                    // TODO Later versions of rust can use `to_bits`
+                    F32Literal(v) => (v as u32).hash(&mut self.hasher),
+                    F64Literal(v) => (v as u64).hash(&mut self.hasher),
+                }
+            }
+            Ident(ref sym) => {
+                // We track symbols to disambiguate redefinitions, but also to ignore the actual
+                // textual symbol name. By hashing a number representing the symbol, expressions
+                // which are structurally the same but have different names will produce the same
+                // hash value.
+                match symbol_positions.entry(sym) {
+                    Entry::Occupied(ref ent) => {
+                        ent.get().hash(&mut self.hasher);
+                    },
+                    _ => {
+                        return weld_err!("Undefined symbol {}", sym);
+                    }
+                }
+            }
+            BinOp {ref kind, .. } => {
+                kind.hash(&mut self.hasher);
```

Comment:
why don't you call `from_expr` on the children of the `BinOp`?

Reply:
At the bottom of this match there's a `for child in expr.children()` that handles the subexpressions - the match part just handles the non-subexpression fields (e.g., the `BinOpKind` or the actual literal value).